plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/SpringReload.java

```diff
@@ -0,0 +1,572 @@
+                beanFactory.configureBean(bean, beanName);
+            }
+        }
+        // 6.3 process @Value and @Autowired of singleton beans excluding destroyed beans
+        AutowiredAnnotationProcessor.processSingletonBeanInjection(beanFactory);
+        // 6.4 reset again
+        ConfigurationClassPostProcessorEnhance.getInstance(beanFactory).postProcess(beanFactory);
+        // 6.5 invoke getBean to instantiate singleton beans
+        preInstantiateSingleton(newBeanNames);
+        // 6.6 reset mvc initialized, it will update the mapping of url and handler
+        ResetRequestMappingCaches.reset(beanFactory);
+        return true;
+    }
+
+    private void preInstantiateSingleton(Set<String> newBeanNames) {
+        for (String beanName : beanFactory.getBeanDefinitionNames()) {
+            // 重新创建bean，并初始化
+            Object singleton = beanFactory.getSingleton(beanName);
+            BeanDefinition beanDefinition = null;
+            try {
+                beanDefinition = beanFactory.getBeanDefinition(beanName);
+            } catch (NoSuchBeanDefinitionException e) {
+                LOGGER.debug("bean not found: " + beanName);
+                continue;
+            }
+            if (beanDefinition.isSingleton()) {
+                Object target = beanFactory.getBean(beanName);
+                if (singleton == null) {
+                    if (newBeanNames.contains(beanName)) {
+                        LOGGER.reload("the bean '{}' of {} is singleton and new created", beanName, ClassUtils.getUserClass(target).getName());
+                    } else {
+                        LOGGER.reload("the bean '{}' of {} is singleton and recreated", beanName, ClassUtils.getUserClass(target).getName());
+                    }
+                }
+            } else if (beanDefinition.isPrototype()) {
+                if (recreatedBeans.contains(beanName)) {
+                    LOGGER.reload("bean '{}' is protoType and reloaded", beanName);
+                }
+            }
+        }
+    }
+
+    private void clearCache() {
+        // spring won't rebuild dependency map if injectionMetadataCache is not cleared
+        // which lead to singletons depend on beans in xml won't be destroy and recreate, may be a spring bug?
+        ResetSpringStaticCaches.reset();
+        ResetBeanPostProcessorCaches.reset(beanFactory);
+        ResetTransactionAttributeCaches.reset(beanFactory);
+        ResetBeanFactoryPostProcessorCaches.reset(beanFactory);
+        ProxyReplacer.clearAllProxies();
+        // fixme temperately disable it
+//        ResetBeanFactoryCaches.reset(beanFactory);
+        ConfigurationClassPostProcessorEnhance.getInstance(beanFactory).postProcess(beanFactory);
+        ResetAnnotationCache.resetAnnotationScanner(beanFactory);
+    }
+
+    private void reloadAnnotatedBeanDefinitions(Class clazz, String[] beanNames, List<String> needReloadBeanDefinitionNames) {
```

Comment:
I think we should pass in a list of classes in the 1st param, and return a list of 'needReloadBeanDefinitonNames' like this:

```java
private List<String findConfigurationClassesToReload(List<Class<?>> classes, String[] beanNames) {

}
```

Reply:
it should update the beanDefinition or remove attribute.